llvm/lib/IR/Value.cpp

```diff
@@ -775,6 +775,25 @@ const Value *Value::stripAndAccumulateConstantOffsets(
           V = RV;
         if (AllowInvariantGroup && Call->isLaunderOrStripInvariantGroup())
           V = Call->getArgOperand(0);
+    } else if (auto *Int2Ptr = dyn_cast<Operator>(V)) {
+      // Try to accumulate across (inttoptr (add (ptrtoint p), off)).
+      if (!Int2Ptr || Int2Ptr->getOpcode() != Instruction::IntToPtr ||
+          Int2Ptr->getOperand(0)->getType()->getScalarSizeInBits() != BitWidth)
+        return V;
+      auto *Add = dyn_cast<AddOperator>(Int2Ptr->getOperand(0));
+      if (!AllowNonInbounds || !Add)
+        return V;
+      auto *Ptr2Int = dyn_cast<PtrToIntOperator>(Add->getOperand(0));
+      auto *CI = dyn_cast<ConstantInt>(Add->getOperand(1));
+      if (!Ptr2Int || !CI)
+        return V;
+
+      APInt AddOffset = CI->getValue();
+      if (AddOffset.getSignificantBits() > BitWidth)
+        return V;
+
+      Offset = AddOffset.sextOrTrunc(BitWidth);
```

Comment:
Didn't you already very above that AddOffset must have bit width BitWidth via the `Int2Ptr->getOperand(0)->getType()->getScalarSizeInBits()` check?